pkg/steps/test_test.go

```diff
@@ -0,0 +1,179 @@
+			},
+			Annotations: map[string]string{
+				"ci.openshift.io/job-spec": "",
+			},
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:                     step.name,
+					Image:                    "somename:sometag",
+					Command:                  []string{"/bin/sh", "-c", "#!/bin/sh\nset -eu\nlaunch-tests"},
+					TerminationMessagePolicy: v1.TerminationMessageFallbackToLogsOnError,
+				},
+			},
+			RestartPolicy: v1.RestartPolicyNever,
+		},
+		Status: v1.PodStatus{Phase: phaseAfterRun},
+	}
+}
+
+func TestPodStepMethods(t *testing.T) {
+	namespace := "TestNamespace"
+	ps, spec, _ := preparePodStep(t, namespace)
+	examineStep(t, ps, spec)
+}
+
+func TestPodStepExecution(t *testing.T) {
+	namespace := "TestNamespace"
+	testCases := []struct {
+		purpose        string
+		podStatus      v1.PodPhase
+		expectRunError bool
+	}{
+		{
+			purpose:        "Pod run by PodStep succeeds so PodStep terminates and returns no error",
+			podStatus:      v1.PodSucceeded,
+			expectRunError: false,
+		}, {
+			purpose:        "Pod run by PodStep fails so PodStep terminates and returns an error",
+			podStatus:      v1.PodFailed,
+			expectRunError: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.purpose, func(t *testing.T) {
+			ps, _, client := preparePodStep(t, namespace)
+			expectedPod := makeExpectedPod(ps, tc.podStatus)
+
+			executionExpectation := executionExpectation{
+				prerun: doneExpectation{
+					value: false,
+					err:   false,
+				},
+				runError: tc.expectRunError,
+				postrun: doneExpectation{
+					value: true,
+					err:   false,
+				},
+			}
+
+			watcher, err := client.Pods(namespace).Watch(meta.ListOptions{})
+			if err != nil {
+				t.Errorf("Failed to create a watcher over pods in namespace")
+			}
+			defer watcher.Stop()
+
+			clusterBehavior := func() {
+				// Expect a single event (a Creation) to happen
+				// Immediately set the Pod status to Succeeded, because
+				// that is what the step waits on
+				for {
+					event, ok := <-watcher.ResultChan()
+					if !ok {
+						t.Log("Fake cluster: watcher event closed, exiting()")
+						break
+					}
+					if pod, ok := event.Object.(*v1.Pod); ok {
+						t.Logf("Fake cluster: Received event on pod '%s': %s", pod.ObjectMeta.Name, event.Type)
+						t.Logf("Fake cluster: Updating pod '%s' status to '%s' and exiting", pod.ObjectMeta.Name, tc.podStatus)
+						// make a copy to avoid a race
+						newPod := *pod
+						newPod.Status.Phase = tc.podStatus
+						if _, err := client.Pods(namespace).UpdateStatus(&newPod); err != nil {
+							t.Errorf("Fake cluster: UpdateStatus() returned an error: %v", err)
+						}
+						break
+					}
+					t.Logf("Fake cluster: Received non-pod event: %v", event)
+				}
+			}
+
+			executeStep(t, ps, executionExpectation, clusterBehavior)
+
+			if pod, err := client.Pods(namespace).Get(ps.Name(), meta.GetOptions{}); !reflect.DeepEqual(expectedPod, pod) {
```

Comment:
When comparing API objects you need `semantic.Equality.DeepEqual()`

Reply:
(in some objects the order of a slice may not be guaranteed or semantically important for instance)

Reply:
Addressed.